Added formula for growing degree days and vernalization

poudelpratishtha opened the pull request.

krsnapaudel commented on util/features.py:
```diff
@@ -112,9 +111,46 @@ def aggregate_by_period(df, index_cols, period_col, aggrs, ft_cols):
+   tmin = np.asarray(tmin)
+   tmax = np.asarray(tmax)
+   
+   tavg = (tmin + tmax) / 2
```
There is already a tavg in data. Do you need to do this?

krsnapaudel reviewed (changes requested):
Two comments:
1. Add other features from table.
2. Integrate with current feature design code. Create a feature column and feature data frame. See `design_features`.

poudelpratishtha commented:
I removed tavg calculation and added ndvi and climatic water balance. A small sample data frame to test the function would help. Let me know if there's any such df.

poudelpratishtha commented:
The tests are failing because of column names. Not sure what the correct column names are.

krsnapaudel commented on util/features.py:
```diff
@@ -218,9 +245,13 @@ def design_features(
     # TODO: 2. Add code to make drainage class a categorical feature.
+
+    if 'DRAIN' in soil_df.columns:
+        soil_df['DRAIN'] = soil_df['DRAIN'].astype('category')
```
The sample data from NL does not yet have "drainage_class". The one now uploaded in Google drive has.

krsnapaudel commented on util/features.py:
```diff
@@ -244,6 +275,18 @@ def design_features(
+        weather_df = weather_df.merge(et0_df, on=index_cols)
+        weather_df['cwb'] = weather_df['prec'] - weather_df['ET0']
+
+    weather_df['cumul_cwb'] = weather_df.groupby(KEY_LOC)['cwb'].cumsum()
```
Weather data from cybench will now contain et0. So existing code will also need an update.

krsnapaudel commented on util/features.py:
```diff
@@ -198,7 +224,7 @@ def unpack_time_series(df, indicators):
 
 def design_features(
-    weather_df, soil_df, fapar_df, ndvi_df=None, et0_df=None, soil_moisture_df=None
+    weather_df, soil_df, fapar_df, crop, ndvi_df=None, et0_df=None, soil_moisture_df=None
```
Crop should be the first argument.

krsnapaudel commented on config.py:
```diff
@@ -33,6 +33,9 @@
 WEATHER_INDICATORS = ["tmin", "tmax", "tavg", "prec"]
 
+# Base temperature for corn and wheat for growing degree days wheat:0 corn:10
+BASE_TEMP = {"wheat": 0, "corn": 10}
```
Better to make it GDD_BASE_TEMP

krsnapaudel commented on util/features.py:
```diff
@@ -218,31 +244,46 @@ def design_features(
     period_length = "month"
-    max_feature_cols = ["fapar"]  # ["ndvi", "fapar"]
-    avg_feature_cols = ["tmin", "tmax"]  # , "tmax", "tavg", "prec", "rad"]
+    max_feature_cols = ["ndvi", "fapar", "cum_gdd", "cum_cwb", "cum_ndvi", "cum_prec"]
```
They need to be separated into 3.
1. weather max
2. fapar max
3. NDVI max
These 3 data sources have to be handled separately. They have different time resolution.
